### DISCUSSION TITLE
[EE6] Improvements to storage and access of Pages module data

### DISCUSSION THREAD

**Discussion by PaulBaileyAtChan** [ExpressionEngine/ExpressionEngine#3310] [Feature Requests]
The Pages module currently stores data in an encoded form, within a single cell of the EE `exp_sites` table. This presents a number of significant issues:

- It is potentially a single point of failure in the case of database corruption.
- The data are not human readable/modifiable within the database itself. This is in contrast to essentially all of the per-entry (and per-template) data stored elsewhere within the database. In the general case, the ability to both view and edit information directly in the database is an important and useful fallback, for situations where the CP option is unavailable, or ineffective (for example, in situations involving DB corruption).
- It scales poorly. The size of the encoded data is proportional to the number of pages present, and all must be decoded/encoded each time, even if only a single entry is needed. This is inefficient, and works against the relational database model.

Without being too specific about how a solution might be implemented:

- Pages module data should be stored in a human readable/modifiable form.
- Data should be stored as discrete per-entry values, to allow efficient access.
- Data should be in their own table, or as a new column in a suitable, existing table which stores data per-entry.

**Comment by robsonsobral** [comment]
There's a legend Pages Module was coded by a guy name Leslie and other developer in a hurry. I don't know for sure if it's true, but is totally believable.

I don't think we should have page modules anymore:

- it should be a per channel configuration. Something like **Allow URI override**;
- ~~the URI could be saved on `channel_data_titles` table~~ the URI should have its own table together with some nesting options;
- a hierarchical data field type would be fucking useful to create menus and keep pages in order. Even better, the Pages Module could be used just for this: put things in order without the need to edit entry by entry;
- a template field type (or to allow templates as source to the Select field type) would close the deal.

Too much?

**Comment by PaulBaileyAtChan** [comment]
I have no idea about the origins of the Pages module, but I've always had it in the back of my mind that part of the reason Ellis Labs always pretty much actively discouraged people from using it, was that they were aware the implementation wasn't very robust.

I worry about it the way that I worried about Macromedia applications when they were taken over by Adobe, who had similar applications they were likely to prioritise (RIP Fireworks). It's understandable that Packet Tide would put more time and effort into Structure, which is more familiar to them, and is directly connected to revenue.

Pages obviously isn't the only tool for mapping URLs to content, and in most cases isn't the right one, but it is still an important part of EE, and, despite the greater attention that Structure gets these days (for obvious reasons), Pages is still the core EE method for assigning URLs to single entries. It's _long overdue_ some love.

My hopes are modest, and I hope realistic: rework how the Pages module data are stored, for robustness and efficiency. If that happens as part of some wider development of the module, great.

(I also have another proposal in Github to improve the CP display of all entries with URLs assigned in Pages, to allow expansion / collapse of the nested list: https://github.com/ExpressionEngine/ExpressionEngine/issues/238 This is much more on the usability front.)

**Comment by robsonsobral** [comment]
I'm thinking more in a broader approach. EECMS should use more EECMS. Why are the category fields different from the channel fields? Remember the member templates? Why sites can't have custom fields? So, I'm thinking of improvements that could be widely used, not just on Pages Module.

Besides that, some addons, the ones which fill gaps, instead of add functionalities, should be incorporated into core for eecms to survive. To cripple it to force the selling of an addon is a proved bad move by the own history of eecms.

Let's dream!
